### ISSUE TITLE
Live example is throwing a null pointer exception in javascript

### ISSUE THREAD

**Issue by page3131** [angular/angular#19602]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
 I go to https://angular.io/generated/live-examples/toh-pt6/eplnkr.html for the live. I try to do anything and it is not responding. Also on load it throws a console log error. So this is to report two errors.
I was just looking through the code and it looks like it is also supposed to be populated with heros.

## Expected behavior
I should be able to add a hero. I tried two different browsers. No Luck. It doesn't get populated with heros to start with.

## Minimal reproduction of the problem with instructions
Pasted https://angular.io/generated/live-examples/toh-pt6/eplnkr.html in a browser. Click on Heroes and enter in a heros name then click on the add button. The name will disappear. I also noticed that on load it throws an error.

## What is the motivation / use case for changing the behavior?
To have good documentation

## Environment

<pre><code>
Angular version: X.Y.Z
This is running on the plnkr.co I have no idea what version it is under.

Browser:
- [x] Chrome (desktop) version 61
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [x] Firefox version 55
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: XX  
- Platform:  

Others:
how about both of the error messages:
INITIAL LOADING ERROR:
ERROR Error: Uncaught (in promise): TypeError: Cannot read property 'slice' of undefined
TypeError: Cannot read property 'slice' of undefined
    at eval (https://run.plnkr.co/preview/cj8glfx1h000r3c5scc0ie3qi/app/dashboard.component.ts!transpiled:22:68)
    at ZoneDelegate.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:242:26)
    at Object.onInvoke (https://unpkg.com/@angular/core/bundles/core.umd.js:3922:33)
    at ZoneDelegate.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:241:32)
    at Zone.run (https://unpkg.com/zone.js@0.7.4/dist/zone.js:113:43)
    at https://unpkg.com/zone.js@0.7.4/dist/zone.js:520:57
    at ZoneDelegate.invokeTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:275:35)
    at Object.onInvokeTask (https://unpkg.com/@angular/core/bundles/core.umd.js:3913:33)
    at ZoneDelegate.invokeTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:274:40)
    at Zone.runTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:151:47)
    at drainMicroTaskQueue (https://unpkg.com/zone.js@0.7.4/dist/zone.js:418:35)
    at ZoneTask.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:349:25)
    at data.args.(anonymous function) (https://unpkg.com/zone.js@0.7.4/dist/zone.js:1224:25)
    at eval (https://run.plnkr.co/preview/cj8glfx1h000r3c5scc0ie3qi/app/dashboard.component.ts!transpiled:22:68)
    at ZoneDelegate.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:242:26)
    at Object.onInvoke (https://unpkg.com/@angular/core/bundles/core.umd.js:3922:33)
    at ZoneDelegate.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:241:32)
    at Zone.run (https://unpkg.com/zone.js@0.7.4/dist/zone.js:113:43)
    at https://unpkg.com/zone.js@0.7.4/dist/zone.js:520:57
    at ZoneDelegate.invokeTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:275:35)
    at Object.onInvokeTask (https://unpkg.com/@angular/core/bundles/core.umd.js:3913:33)
    at ZoneDelegate.invokeTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:274:40)
    at Zone.runTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:151:47)
    at drainMicroTaskQueue (https://unpkg.com/zone.js@0.7.4/dist/zone.js:418:35)
    at ZoneTask.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:349:25)
    at data.args.(anonymous function) (https://unpkg.com/zone.js@0.7.4/dist/zone.js:1224:25)
    at resolvePromise (https://unpkg.com/zone.js@0.7.4/dist/zone.js:486:31) [angular]
    at https://unpkg.com/zone.js@0.7.4/dist/zone.js:523:17 [angular]
    at Object.onInvokeTask (https://unpkg.com/@angular/core/bundles/core.umd.js:3913:33) [angular]
    at ZoneDelegate.invokeTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:274:40) [angular]
    at Zone.runTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:151:47) [<root> => angular]
    at drainMicroTaskQueue (https://unpkg.com/zone.js@0.7.4/dist/zone.js:418:35) [<root>]
    at ZoneTask.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:349:25) [<root>]
    at data.args.(anonymous function) (https://unpkg.com/zone.js@0.7.4/dist/zone.js:1224:25) [<root>]

ADD HERO ERROR:
ERROR Error: Uncaught (in promise): TypeError: Cannot read property 'push' of undefined
TypeError: Cannot read property 'push' of undefined
    at eval (https://run.plnkr.co/preview/cj8glfx1h000r3c5scc0ie3qi/app/heroes.component.ts!transpiled:34:26)
    at ZoneDelegate.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:242:26)
    at Object.onInvoke (https://unpkg.com/@angular/core/bundles/core.umd.js:3922:33)
    at ZoneDelegate.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:241:32)
    at Zone.run (https://unpkg.com/zone.js@0.7.4/dist/zone.js:113:43)
    at https://unpkg.com/zone.js@0.7.4/dist/zone.js:520:57
    at ZoneDelegate.invokeTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:275:35)
    at Object.onInvokeTask (https://unpkg.com/@angular/core/bundles/core.umd.js:3913:33)
    at ZoneDelegate.invokeTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:274:40)
    at Zone.runTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:151:47)
    at drainMicroTaskQueue (https://unpkg.com/zone.js@0.7.4/dist/zone.js:418:35)
    at ZoneTask.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:349:25)
    at data.args.(anonymous function) (https://unpkg.com/zone.js@0.7.4/dist/zone.js:1224:25)
    at eval (https://run.plnkr.co/preview/cj8glfx1h000r3c5scc0ie3qi/app/heroes.component.ts!transpiled:34:26)
    at ZoneDelegate.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:242:26)
    at Object.onInvoke (https://unpkg.com/@angular/core/bundles/core.umd.js:3922:33)
    at ZoneDelegate.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:241:32)
    at Zone.run (https://unpkg.com/zone.js@0.7.4/dist/zone.js:113:43)
    at https://unpkg.com/zone.js@0.7.4/dist/zone.js:520:57
    at ZoneDelegate.invokeTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:275:35)
    at Object.onInvokeTask (https://unpkg.com/@angular/core/bundles/core.umd.js:3913:33)
    at ZoneDelegate.invokeTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:274:40)
    at Zone.runTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:151:47)
    at drainMicroTaskQueue (https://unpkg.com/zone.js@0.7.4/dist/zone.js:418:35)
    at ZoneTask.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:349:25)
    at data.args.(anonymous function) (https://unpkg.com/zone.js@0.7.4/dist/zone.js:1224:25)
    at resolvePromise (https://unpkg.com/zone.js@0.7.4/dist/zone.js:486:31) [angular]
    at https://unpkg.com/zone.js@0.7.4/dist/zone.js:523:17 [angular]
    at Object.onInvokeTask (https://unpkg.com/@angular/core/bundles/core.umd.js:3913:33) [angular]
    at ZoneDelegate.invokeTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:274:40) [angular]
    at Zone.runTask (https://unpkg.com/zone.js@0.7.4/dist/zone.js:151:47) [<root> => angular]
    at drainMicroTaskQueue (https://unpkg.com/zone.js@0.7.4/dist/zone.js:418:35) [<root>]
    at ZoneTask.invoke (https://unpkg.com/zone.js@0.7.4/dist/zone.js:349:25) [<root>]
    at data.args.(anonymous function) (https://unpkg.com/zone.js@0.7.4/dist/zone.js:1224:25) [<root>]
</code></pre>

**Comment by gkalpak** [comment]
It seems that in `HeroService` there are some occurrences of `.then(response => response.json().data)` which should actually be `.then(response => response.json())`.

@wardbell, any idea what is going on here? Is this related to the recent `angular-in-memory-web-api` update?

**Comment by page3131** [comment]
No idea. I hardly have a clue about angular this is the first. I was hoping
some one could fix it. but it looks like the examples later on seem to work.

**Comment by kenji-1996** [comment]
@gkalpak THANK YOU SO MUCH
Im so glad issues are reported and replied to so damn quickly, i was stuck on the tutorial and im glad I could reach an answer to continue my learning!

**Comment by gkalpak** [comment]
To add a little more context:

The reason the example fails on plnkr is that `angular-in-memory-web-api` was [recently updated][1] to drop the `.data` property in returned mock responses (by default). So, one should either remove all `.data` occurences from returned responses or configure `angular-in-memory-web-api` to add the `.data` property.

The examples (and the associated tests) are not failing on our builds, because we have the `angular-in-memory-web-api` version pinned to `0.4.x` (before the change that removes `.data` by default).

https://github.com/angular/angular/blob/d7eac7ee56fef639a74327ee2001f9455d2cf615/aio/tools/examples/shared/package.json#L28

The generated plnkrs, on the other side, always use the latest stable version available on npm (currently 0.5.x).

##
While we are investigating how to best address this, the easiest/safest option (imo) is to pin the version on the plnkr to 0.4.x. I.e. manually make the following change in your plnkrs `system.config.js`:

```diff
-'angular-in-memory-web-api': 'npm:angular-in-memory-web-api/bundles/in-memory-web-api.umd.js',
+'angular-in-memory-web-api': 'npm:angular-in-memory-web-api@0.4/bundles/in-memory-web-api.umd.js',
```

[1]: https://github.com/angular/in-memory-web-api#important-recent-changes

**Comment by keego** [comment]
Just ran into this same issue while following the Tour of Heroes tutorial. I imagine most people following the tutorial won't have the `angular-in-memory-web-api` pinned as you guys do. Would be great to have a little sub-section noting that the reader may need to drop the `.data` and just use `.json()` instead depending on the version of their `angular-in-memory-web-api`.

**Comment by kenji-1996** [comment]
Really messes with new people like me building knowledge of a new framework

In fact I've had so many issues with small interactions with the official angular and angular material

I'd imagine it would scare some new devs away

On the bright side all theses issues I'm having allow me to overcome then by learning and interacting with the community here and on stack overflow

**Comment by GaboSantyCosta** [comment]
**At least two things could contribute.**

1) If you are working with firebase to do the example you should note that this line of code: **.then (response => response.json().data as Hero [])** should be **.then (response => response.json() as Hero [])** ".data" must be deleted.

2) This line of code: "private heroesUrl = 'api / heroes';" would not apply to the case since if we decide to use "https://heroes-xXxxX.firebaseio.com/heores" we must add the extension of the data to obtain **"https://heroes-xXxxX.firebaseio.com/heores/. json"** otherwise we will get **"CORS 'Access-Control-Allow-Origin' ERROR"**. In the case of getting a single hero it should be instead of this: "const url =` $ {this.heroesUrl} / $ {id} `" this **"` $ {'https://heroes-xXxxX.firebaseio.com / heores'} / $ {id} .json` "** which should apply to the other examples.

Sorry my english... google translate. xD

**Comment by gkalpak** [comment]
Fixed in #19668.

**Comment by pavelmorozov** [comment]
I am not sure. This marked fixed, though same issue with me.

When I run code example it contains
`response.json().data as Hero[];`
and it works

And when I work with manual and create local application it still contains
`response.json().data as Hero[]`
same string, though data should not be there or error happen

I tried to set in package.json
`"angular-in-memory-web-api": "~0.4.0"`
and see in debugger data property on response json is absent. 

So what is the fix? Come back to older angular-in-memory-web-api or delete .data property call? What route should I take to finish with Tour of heroes?

**Comment by gkalpak** [comment]
This is fixed and should work both for the [live example][1] and the [downloadable zip][2].
If you are re-creating the example app "by hand" make sure your end results looks the same as the zip. If the problem persists, please post exact reproduction steps, so we can investigate.

[1]: https://angular.io/generated/live-examples/toh-pt6/eplnkr.html
[2]: https://angular.io/generated/zips/toh-pt6/toh-pt6.zip

**Comment by pavelmorozov** [comment]
Right it works if I download zip file. There i see in hero.service.ts
`response.json().data as Hero[];`
and in package.json:
`"angular-in-memory-web-api": "~0.4.0",`

Though when you create application by hand, it not works. This because you have to add some files dependent on angular-in-memory-web-api, like in-memory-data.service.ts. If you run application you will  get an error like:
` ...Cannot find module 'angular-in-memory-web-api'.`
I not see this error description in manual, and after some search I tried 
```
npm install --save angular-in-memory-web-api
my-app@0.0.0 /home/pm/Documents/my-app
└── angular-in-memory-web-api@0.5.1 
```
as you can see 0.5.1 instead of 0.4.0, and in my package json, I see now the 
`"angular-in-memory-web-api": "^0.5.1",`

This leads "application made by hand" not works, so you get in debug and see the .data property is absent in response. I know this is simple thing for anybody who works with latest Angular, though it is not for anybody who tries to use step-by-step manual.

If I will just change package.json to
`"angular-in-memory-web-api": "~0.4.0",`
and restart, it is still not works. The data property not exist in response. This because the newer version still in node_modules. To solve I deleted  node_modules folder and run 
`npm install`
now application works. So it looks like manual page should be updated with info about angular-in-memory-web-api setup and version.

**Comment by gkalpak** [comment]
I don't follow most of what you are saying (it's probably me, not you :grin:)
The dependency on `angular-in-memory-web-api` and the related files (such as `in-memory-data.service.ts`) are clearly mentioned in the guide. Also, updating `package.json` does not automatically update the installed dependencies. You have to `npm install` again. This is not Angular-related though; it is how node/npm works.

Version compatibility is indeed an issue, but since versions change, I would avoid hard-coding them inside the guide itself, becaue they will inevitably become outdated. You can always check the provided live or downloadable example and see the versions/ranges used there. And you can also compare your "hand-written version with the provided full example and see where you deviated.

Finally, such breaking changes are not common in our examples dependencies and we will hopefully upgrade `angular-in-memory-web-api` to latest version soon anyway.

**Comment by pavelmorozov** [comment]
Thanks for comment. It makes it clear. Any way it is up to you (your team) how to maintain the manuals. :)

**Comment by gbgames** [comment]
Two changes I had to make that were not in the tutorial:
1) Thanks to finding https://stackoverflow.com/a/46789415 I learned that the in memory database needed to be installed. "As of October 17, 2017, the tutorial fails to mention that angular-in-memory-web-api is not included in a standard CLI build, and must be installed separately. This can be easily done with npm:

$ npm install angular-in-memory-web-api --save

This automatically handles the necessary changes to package.json, so you don't need to make edits yourself."

2) Thanks to this issue and GaboSantyCosta's post in particular, I learned that the .data needs to be removed in that one line. Otherwise, the tutorial says "Refresh the browser. The hero data should successfully load from the mock server." when in fact it looks like the hero data is missing.
